### ISSUE TITLE
API to convert workflow format to prompt format

### ISSUE THREAD

**Issue by jamesWalker55** [comfyanonymous/ComfyUI#1112]
When using the API, it uses a different format than workflow JSON files. This format can be retrieved from EXIF data in saved images, however it can be inconvenient to 1) build a workflow, then 2) generate an image, 3) load the EXIF data from the image, finally 4) use the new prompt in the API.

I would like to suggest an API endpoint that receives a workflow JSON object, converts it to a prompt and returns it. This way, the whole process can be easily automated with a Python script by making a request to that endpoint, then using the returned JSON to make API requests.

**Comment by pydn** [comment]
Click the settings wheel, then "enable dev mode options" in the pop up. This enables a save (API Format) button.

**Comment by guinoise** [comment]
This functionnality is required for the sake of simplicity in api consumer. 

To use correctly the prompt api we have to send: 
1. client_id: this is not an issue
2. prompt : the api json
3. extra_data: { extra_pnginfo: { workflow json }

The third point is required to:
1. Embed the workflow in the generated images
2. Some nodes require this information to work

As of today, this mean you have to export the workflow.json and the workflow_api.json, then do the changes in BOTH files and send them to the api. 

If a functionality to convert the workflow to workflow_api or even better an option to push the workflow.json to the prompt API exists, it will make the life of the API consumers a lot easier.

**Comment by IntenseSunlight** [comment]
I used this simple, hacky way to get the full prompt for the api.  I agree there should be a better way to get this with the normal "Save (api_format)" operation.

In `./server.py`, line 450+ at the time of this writing
```python
        .
        .
        @routes.post("/prompt")
        async def post_prompt(request):
            print("got prompt")
            resp_code = 200
            out_string = ""
            json_data =  await request.json()
            json_data = self.trigger_on_prompt(json_data)
            with open("last_prompt.json", "w") as f:              # <- dump prompt to file
                json.dump(json_data, f, indent=4)                 #  in local directory

            if "number" in json_data:
            .
            . (etc.)
```

**Comment by dmarx** [comment]
looks like the `graphToPrompt()` function is probably the conversion code we're all looking for.

https://github.com/comfyanonymous/ComfyUI/blob/master/web/scripts/app.js#L1857-L1989

**Comment by imperator-maximus** [comment]
we have a large external application and do a lot of ComfyUI graph generation there and we have to use the normal format. But now we need to execute the workflow of course which is impossible by that internal API format without doing that same conversion on our side.
So I suggest a refactoring doing that conversion only on server-side in Python so both formats could be accepted there.

**Comment by RyoJerryYu** [comment]
> we have a large external application and do a lot of ComfyUI graph generation there and we have to use the normal format. But now we need to execute the workflow of course which is impossible by that internal API format without doing that same conversion on our side. So I suggest a refactoring doing that conversion only on server-side in Python so both formats could be accepted there.


For those coming from search engines (like me): Since the merge of https://github.com/comfyanonymous/ComfyUI/pull/4379 , ComfyUI's main branch has removed the original frontend code. The frontend code referenced here refers to: https://github.com/comfyanonymous/ComfyUI/blob/951a2064a34e3e2ab468942663fa59df6a212af3/web/scripts/app.js#L1857-L1989

BTW, I'm still unable to find a program-friendly way to convert from `workflow.json` to `workflow_api.json` , except that implement the `graphToPrompt()` in my code. I'm really looking forward to seeing this feature implemented on the backend of ComfyUI.

**Comment by oumad** [comment]
> > looks like the `graphToPrompt()` function is probably the conversion code we're all looking for.
> > https://github.com/comfyanonymous/ComfyUI/blob/master/web/scripts/app.js#L1857-L1989
> 
> For those coming from search engines (like me): Since the merge of #4379 , ComfyUI's main branch has removed the original frontend code. The frontend code referenced here refers to:
> 
> https://github.com/comfyanonymous/ComfyUI/blob/951a2064a34e3e2ab468942663fa59df6a212af3/web/scripts/app.js#L1857-L1989
> 
> BTW, I'm still unable to find a program-friendly way to convert from `workflow.json` to `workflow_api.json` , except that implement the `graphToPrompt()` in my code. I'm really looking forward to seeing this feature implemented on the backend of ComfyUI.

@RyoJerryYu Can you share how you implemented that in your code ? my attempts have failed at producing a correct workflow_api.json file, I have imported the litegraph library and copied the graphToPrompt function to use in my app in this manner : 

```js
import { LGraph, LGraphEventMode } from '@comfyorg/litegraph';

async function processData(graphData, clean = true) {

  graph = new LGraph()
  graph.start()
  graph.configure(graphData)
  
  // rest of graphToPrompt code without "this."
  // ...
}
```

I don't know what I'm missing

**Comment by RyoJerryYu** [comment]
> @RyoJerryYu Can you share how you implemented that in your code ? my attempts have failed at producing a correct workflow_api.json file, I have imported the litegraph library and copied the graphToPrompt function to use in my app in this manner :
> 
> ```js
> import { LGraph, LGraphEventMode } from '@comfyorg/litegraph';
> 
> async function processData(graphData, clean = true) {
> 
>   graph = new LGraph()
>   graph.start()
>   graph.configure(graphData)
>   
>   // rest of graphToPrompt code without "this."
>   // ...
> }
> ```
> 
> I don't know what I'm missing. By the way the referenced code has moved to the comfyUI front end project here : https://github.com/Comfy-Org/ComfyUI_frontend/blob/b0cc5eac959877fbc72acbb80cbe437c9ada5050/src/scripts/app.ts#L2303

Actually, I didn't manage to implement it successfully. It seems to rely heavily on frontend code from custom nodes, and implementing it ourselves would require a significant amount of work. In the end, we chose to modify the main flow of our project, so that it is no longer necessary to convert from workflow.json to workflow_api.json in our backend.

**Comment by piewared** [comment]
Agreed, these would be very useful. The administrative burden of having to maintain synchronized copies of workflow.json and workflow_api.json files for a large scale project is unfortunate — especially when converting between the two formats requires a manual step.

**Comment by destlaver** [comment]
this is still a thing. API are unusable. What's the point of having an api if you have to do 1 step manually. It's non-sense

**Comment by saurabhthesuperhero** [comment]
@destlaver no success yet right? I managed to conver just handful of nodes, but thats not good or reliable.

**Comment by llyyfree** [comment]
+1

**Comment by tiantianaixuexi** [comment]
I need this too. When I use export workflow directly, its widget_values ​​are hard to parse. For example, the KSampler node will have an extra ComfyUI editor page-specific data "randomize". It would be great if I could use the API to export workflow (API).

**Comment by Ccheers** [comment]
same here

**Comment by Fernandosa21** [comment]
Hope this helps anyone

I found this custom node https://github.com/ShunL12324/comfy-portal-endpoint that has an API to get and convert a workflow to prompt format. The only issue I found is that for it to work properly, ComfyUI needs to be opened in a browser; if Comfy is not opened in any browser, the API responds with a timeout.

**Comment by SethRobinson** [comment]
I had the same issue (my [tools](https://github.com/SethRobinson/aitools_client) need it) and solved the problem by (sort of) porting the javascript workflow API conversion logic into Python so it can just be a standard ComfyUI endpoint.

I put it up if anybody wants to try it.  It's not tested much but it works fine for my stuff:  https://github.com/SethRobinson/comfyui-workflow-to-api-converter-endpoint

**Comment by onzag** [comment]
> I had the same issue (my [tools](https://github.com/SethRobinson/aitools_client) need it) and solved the problem by (sort of) porting the javascript workflow API conversion logic into Python so it can just be a standard ComfyUI endpoint.
> 
> I put it up if anybody wants to try it. It's not tested much but it works fine for my stuff: https://github.com/SethRobinson/comfyui-workflow-to-api-converter-endpoint

I am going to use that I think, there's no much choice to be fair

By the way you can add http endpoints with custom nodes, no need to modify server.py I'd reckon those instructions are not too great to mess with server.py I myself have even open two websocket connections within a custom node because I am hacking the graph, and ironically even within comfyUI I cannot get the damn thing to give me the graph to send to itself, because the javascript client is handling that.

https://github.com/wywywywy/ComfyUI-pause/blob/main/PauseWorkflowNode.py

Look at this.

I however would personally recommend to keep this repository as it is, and have it as a custom node in a separate repository; the reason for that is that you may not be looking for an endpoint in comfyUI itself, you may be looking to convert your workflow to API and then send that to comfyUI; so they serve two separate purposes, that saves a request, you convert it in your client; and then send it to the server, your client just happens to be a python client.

**Comment by SethRobinson** [comment]
You're right!  This can be a custom node!  Thanks.  I went ahead and just updated [my original repo](https://github.com/SethRobinson/comfyui-workflow-to-api-converter-endpoint), I don't see any reason to keep the old version around.  (it can't run except within comfyui anyway)

**Comment by onzag** [comment]
> You're right! This can be a custom node! Thanks. I went ahead and just updated [my original repo](https://github.com/SethRobinson/comfyui-workflow-to-api-converter-endpoint), I don't see any reason to keep the old version around. (it can't run except within comfyui anyway)

by the way...

https://github.com/otavanopisto/ComfyUI-aihub-workflow-exposer/blob/main/js/export-workflow.js

This shows very advanced exporting, I've hacked comfyui server badly; now I have comfyui workflows in gimp working and other clients, this is the server version.

This was part of an education project, likely it is going to be shutdown :( appreciate a star if you find the stuff in this repo useful in the hopes they dont shut it down.

**Comment by Proryanator** [comment]
> +1
> 
> This would be very helpful to our project to have an API endpoint to get the workflow in API format, in order to automate the update of workflows.
> 
> We have built an application that only interacts with ComfyUI through the API, so currently to update a workflow, we must manually export it in API format, upload the file to our git repo, and build a new version of our application. We can download a workflow.json file using the ComfyUI API, but not in api format. The conversion to the api format does not seem trivial, so it would greatly appreciated if there was an endpoint to get workflow in api format.

I can't for the life of me get the `userdata` endpoint to give me the workflow.json file for a given workflow. I can get the list of .json files via:

```shell
"http://0.0.0.0:8188/api/userdata?dir=workflows&recurse=true&full_info=true"
```

But making any sort of subsequent call to get the file gives me an empty array which isn't very useful.

<img width="641" height="145" alt="Image" src="https://github.com/user-attachments/assets/4527e824-130d-4028-a6f6-547ba6e97e16" />
<img width="542" height="139" alt="Image" src="https://github.com/user-attachments/assets/9f5dd3ed-42a8-4a72-be87-50d658626d2c" />
<img width="626" height="164" alt="Image" src="https://github.com/user-attachments/assets/72fcb1e9-fa5f-441b-8846-d4471909a04f" />

How are you able to download the workflow.json file via an API call?

**Comment by atomicitytech** [comment]
I cannot understand how there aren't way more upvotes on this issue. Seth Robinson's extension is the only sane way to use the workflow file as a source-of-truth that can be parameterized or programmatically interacted with.

The native Comfy architecture of depending on manual, human-driven client-agent code for data transform is a true anti-pattern. 

It would be a massive improvement if the comfyui core switched to depending on Seth's API endpoint as the conversion method, or implemented a similar server-side solution and drop the JS code altogether.

ComfyUi-Blender had to take a dependency on Seth's extension to enable smooth user experience... https://github.com/SethRobinson/comfyui-workflow-to-api-converter-endpoint/issues/4

**Comment by chyueyi** [comment]
For anyone still looking for a backend-side solution, we recently open-sourced ComfyUI-Magicframe:

https://github.com/chyueyi/ComfyUI-Magicframe

Our goal was to avoid maintaining separate UI-workflow and API-workflow files. A workflow can be saved as a ComfyUI App, its public inputs/outputs become a readable schema, and the plugin builds the executable prompt graph when the App is called.

This also avoids exposing raw node IDs to API consumers.
